Updated all dependencies, replaced deprecated v1 OutputPin

jonahbron opened the pull request:
The v1 OutputPin trait has been deprecated because it lacks the ability to return an error.  This update replaces it with the v2 OutputPin, and uses an `enum` type to enable the w5500 driver to bubble either a SPI-related error or an OutputPin-related error.

kellerkindt commented on src/lib.rs:
```diff
@@ -101,12 +107,14 @@ pub enum ArpResponses {
 pub struct UdpSocket(Socket);
 
-pub struct W5500<'a, ChipSelect: OutputPin> {
+pub struct W5500<'a, ChipSelectError, ChipSelect: OutputPin<Error = ChipSelectError>> {
```
```rust
pub struct W5500<'a, ChipSelect: OutputPin>
```

kellerkindt commented on src/lib.rs:
```diff
@@ -101,12 +107,14 @@ pub enum ArpResponses {
 
-impl<'b, 'a: 'b, ChipSelect: OutputPin> W5500<'a, ChipSelect> {
+impl<'b, 'a: 'b, ChipSelectError, ChipSelect: OutputPin<Error = ChipSelectError>>
+    W5500<'a, ChipSelectError, ChipSelect>
```
```rust
    W5500<'a, ChipSelect>
    where ChipSelect::Error = ChipSelectError
```

jonahbron replied:
I don't think it likes that:
```
error: equality constraints are not yet supported in where clauses (see #20041)
   --> src/lib.rs:116:33
    |
116 |     W5500<'a, ChipSelect> where ChipSelect::Error = ChipSelectError
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

kellerkindt commented on src/lib.rs:
```diff
@@ -146,14 +154,31 @@ impl<'b, 'a: 'b, ChipSelect: OutputPin> W5500<'a, ChipSelect> {
+        ChipSelectError,
+        ChipSelect: OutputPin<Error = ChipSelectError>,
+        Spi: FullDuplex<u8, Error = SpiError>,
+    > ActiveW5500<'_, '_, '_, SpiError, ChipSelectError, ChipSelect, Spi>
```
```rust
pub struct ActiveW5500<
    'a,
    'b: 'a,
    'c,
    SpiError,
    ChipSelect: OutputPin,
    Spi: FullDuplex<u8, Error = SpiError>,
>(&'a mut W5500<'b, ChipSelect>, &'c mut Spi);

impl<
        SpiError,
        ChipSelectError,
        ChipSelect: OutputPin,
        Spi: FullDuplex<u8, Error = SpiError>,
    > ActiveW5500<'_, '_, '_, SpiError, ChipSelect, Spi>
    where ChipSelect::Error = ChipSelectError
```

jonahbron commented:
Thanks for your feedback.  While the compiler didn't like the ChipSelect::Error = ChipSelectError (see the error I commented in the review above), you're absolutely right, those generics did not need to exist on the structs.  They're now removed.  In fact, I believe the SpiError generic doesn't need to exist either, so I removed it too.  Please review my changes and tell me what you think.

kellerkindt commented:
Hey, you were faster than me! I was just about to tell you the same :)

kellerkindt commented on src/lib.rs:
```diff
@@ -143,17 +151,27 @@ impl<'b, 'a: 'b, ChipSelect: OutputPin> W5500<'a, ChipSelect> {
         &'b mut self,
         spi: &'c mut Spi,
-    ) -> Result<ActiveW5500<'b, 'a, 'c, E, ChipSelect, Spi>, E> {
+    ) -> Result<ActiveW5500<'b, 'a, 'c, ChipSelect, Spi>, TransferError<SpiError, ChipSelectError>>
```
as well as here
```rust
    pub fn activate<'c, Spi: FullDuplex<u8>>(
        &'b mut self,
        spi: &'c mut Spi,
    ) -> Result<
        ActiveW5500<'b, 'a, 'c, ChipSelect, Spi>,
        TransferError<Spi::Error, ChipSelectError>,
    > {
```

kellerkindt commented on src/lib.rs:
```diff
@@ -106,22 +112,24 @@ pub struct W5500<'a, ChipSelect: OutputPin> {
         mode: ConnectionType,
         arp: ArpResponses,
-    ) -> Result<Self, E> {
+    ) -> Result<Self, TransferError<SpiError, ChipSelectError>> {
```
Here we can even skip the SpiError in the function definition!
```rust
    pub fn with_initialisation<'c, Spi: FullDuplex<u8>>(
        chip_select: &'a mut ChipSelect,
        spi: &'c mut Spi,
        wol: OnWakeOnLan,
        ping: OnPingRequest,
        mode: ConnectionType,
        arp: ArpResponses,
    ) -> Result<Self, TransferError<Spi::Error, ChipSelectError>> {
```

kellerkindt commented on src/lib.rs:
```diff
@@ -341,22 +426,35 @@ impl<E, ChipSelect: OutputPin, Spi: FullDuplex<u8, Error = E>> IntoUdpSocket<Uni
+    fn receive(
+        &mut self,
+        destination: &mut [u8],
+    ) -> Result<Option<(IpAddress, u16, usize)>, TransferError<SpiError, ChipSelectError>> {
```
and here
```rust
    ) -> Result<Option<(IpAddress, u16, usize)>, TransferError<Spi::Error, ChipSelect::Error>> {
```

kellerkindt commented on src/lib.rs:
```diff
@@ -341,22 +426,35 @@ impl<E, ChipSelect: OutputPin, Spi: FullDuplex<u8, Error = E>> IntoUdpSocket<Uni
+        ChipSelect: OutputPin<Error = ChipSelectError>,
+        Spi: FullDuplex<u8, Error = SpiError>,
+    > Udp<SpiError, ChipSelectError>
+    for (&mut ActiveW5500<'_, '_, '_, ChipSelect, Spi>, &UdpSocket)
```
and here
```rust
impl<ChipSelect: OutputPin, Spi: FullDuplex<u8>>
    Udp<Spi::Error, ChipSelect::Error>
    for (&mut ActiveW5500<'_, '_, '_, ChipSelect, Spi>, &UdpSocket)
{
```

kellerkindt commented on src/lib.rs:
```diff
@@ -412,7 +510,7 @@ impl<E, ChipSelect: OutputPin, Spi: FullDuplex<u8, Error = E>> Udp<E> for (&mut
         host_port: u16,
         data: &[u8],
-    ) -> Result<(), E> {
+    ) -> Result<(), TransferError<SpiError, ChipSelectError>> {
```
and here (finally)
```rust
    ) -> Result<(), TransferError<Spi::Error, ChipSelect::Error>> {
```

kellerkindt commented:
Thanks!
